pkg/controller/healthcheck/general/deployment.go

```diff
@@ -0,0 +1,116 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package general
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/gardener/gardener-extensions/pkg/controller/healthcheck"
+
+	"github.com/gardener/gardener/pkg/utils/kubernetes/health"
+	"github.com/go-logr/logr"
+	v1 "k8s.io/api/apps/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// DeploymentHealthChecker contains all the information for the Deployment HealthCheck
+type DeploymentHealthChecker struct {
+	logger         logr.Logger
+	seedClient     client.Client
+	shootClient    client.Client
+	deploymentName string
+	checkType      DeploymentCheckType
+}
+
+// DeploymentCheckType in which cluster the check will be executed
+type DeploymentCheckType string
+
+const (
+	DeploymentCheckTypeSeed  DeploymentCheckType = "Seed"
+	DeploymentCheckTypeShoot DeploymentCheckType = "Shoot"
+)
+
+// CheckSeedDeployment is a healthCheck function to check Deployments in the Seed cluster
+func CheckSeedDeployment(deploymentName string) healthcheck.HealthCheck {
+	return &DeploymentHealthChecker{
+		deploymentName: deploymentName,
+		checkType:      DeploymentCheckTypeSeed,
+	}
+}
+
+// CheckSeedDeployment is a healthCheck function to check Deployments in the Shoot cluster
+func CheckShootDeployment(deploymentName string) healthcheck.HealthCheck {
+	return &DeploymentHealthChecker{
+		deploymentName: deploymentName,
+		checkType:      DeploymentCheckTypeShoot,
+	}
+}
+
+// InjectSeedClient injects the seed client
+func (healthChecker *DeploymentHealthChecker) InjectSeedClient(seedClient client.Client) {
+	healthChecker.seedClient = seedClient
+}
+
+// InjectShootClient injects the shoot client
+func (healthChecker *DeploymentHealthChecker) InjectShootClient(shootClient client.Client) {
+	healthChecker.shootClient = shootClient
+}
+
+// SetLoggerSuffix injects the logger
+func (healthChecker *DeploymentHealthChecker) SetLoggerSuffix(provider, extension string) {
+	healthChecker.logger = log.Log.WithName(fmt.Sprintf("%s-%s-healthcheck-deployment", provider, extension))
+}
+
+// Check executes the health check
+func (healthChecker *DeploymentHealthChecker) Check(ctx context.Context, request types.NamespacedName) (*healthcheck.SingleCheckResult, error) {
+	deployment := &v1.Deployment{}
+
+	var err error
+	if healthChecker.checkType == DeploymentCheckTypeSeed {
```

Comment:
Can't we just accept the `client` as func arg?

Reply:
this  would mean that the user has to pass in the client. As we only have two options, I do not really see a reason to do that, but would rather hide that 'complexity'

Reply:
this would mean that the user has to pass in the client. As we only have two options, I do not really see a reason to do that, but would rather hide that 'complexity